File: 1094_car_pooling/car_pooling.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public boolean carPooling(int[][] trips, int capacity) {
        int[] buckets = new int[1001];
        for (int[] trip : trips) {
            buckets[trip[1]] += trip[0];
            buckets[trip[2]] -= trip[0];
        }
        int passengers = 0;
        for (int bucket : buckets) {
            passengers += bucket;
            if (passengers > capacity) {
                return false;
            }
        }
        return true;
    }
}

File: 1094_car_pooling/car_pooling.py
from typing import List


def car_pooling(trips: List[List[int]], capacity: int) -> bool:
    """
    Time: O(n)
    Space: O(1)
    """
    buckets = [0] * 1001
    for p, start, end in trips:
        buckets[start] += p
        buckets[end] -= p
    passengers = 0
    for bucket in buckets:
        passengers += bucket
        if passengers > capacity:
            return False
    return True
